File: python/math/combinatorics.py
MOD = 10**9 + 7


class Comb:
    def __init__(self, n, mod=MOD):
        self.mod = mod
        self.fact = [1] * (n + 1)
        self.ifact = [1] * (n + 1)
        for i in range(1, n + 1):
            self.fact[i] = self.fact[i - 1] * i % mod
        self.ifact[n] = pow(self.fact[n], mod - 2, mod)
        for i in range(n, 0, -1):
            self.ifact[i - 1] = self.ifact[i] * i % mod

    def ncr(self, n, r):
        if r < 0 or r > n:
            return 0
        return self.fact[n] * self.ifact[r] % self.mod * self.ifact[n - r] % self.mod

File: cpp/math/combinatorics.cpp
#include <bits/stdc++.h>
using namespace std;
template <class Mint> struct Comb {
  vector<Mint> fact, ifact;
  Comb(int n = 0) {
    init(n);
  }
  void init(int n) {
    fact.assign(n + 1, 1);
    ifact.assign(n + 1, 1);
    for (int i = 1; i <= n; i++)
      fact[i] = fact[i - 1] * i;
    ifact[n] = inv(fact[n]);
    for (int i = n; i; i--)
      ifact[i - 1] = ifact[i] * i;
  }
  Mint C(int n, int k) {
    return k < 0 || k > n ? 0 : fact[n] * ifact[k] * ifact[n - k];
  }
  Mint P(int n, int k) {
    return k < 0 || k > n ? 0 : fact[n] * ifact[n - k];
  }
};
